File: lexer/src/Lexer/Builder.cpp
#include "Builder.h"
#include "Machines.h"

namespace Lexer {
auto Builder::build() -> Instance<TokenType> {
  auto lexer = std::make_unique<Lexer::Proto<TokenType>>(1024);

  // 注册关键字状态机
  for (auto [keyword, type] :
       std::initializer_list<std::pair<std::string_view, TokenType>>{
         {"var", TokenType::Var},       {"if", TokenType::If},
         {"else", TokenType::Else},     {"elif", TokenType::Elif},
         {"while", TokenType::While},   {"for", TokenType::For},
         {"return", TokenType::Return}, {"in", TokenType::In},
         {"yield", TokenType::Yield},   {"true", TokenType::True},
         {"false", TokenType::False},   {"null", TokenType::Null},
         {"break", TokenType::Break},   {"continue", TokenType::Continue},
         {"struct", TokenType::Struct}, {"interface", TokenType::Interface},
         {"import", TokenType::Import}, {"as", TokenType::As},
         {"from", TokenType::From},     {"pass", TokenType::Pass},
         {"and", TokenType::And},       {"or", TokenType::Or},
         {"not", TokenType::Not},       {"async", TokenType::Async},
         {"await", TokenType::Await},
       }) {
    lexer->register_machine(Machines::create_keyword_machine(keyword, type));
  }

  // 注册字面量状态机
  lexer->register_machine(Machines::create_string_machine());
  lexer->register_machine(Machines::create_integer_machine());

  /*--- 双字符符号---*/
  for (auto [symbol, type] :
       std::initializer_list<std::pair<std::string_view, TokenType>>{
         {"==", TokenType::DoubleEqual},
         {"!=", TokenType::ExclamationEqual},
         {">=", TokenType::GreaterThanEqual},
         {"<=", TokenType::LessThanEqual},
       }) {
    lexer->register_machine(
      Machines::create_double_symbol_machine(symbol, type)
    );
  }

  /*--- 单字符符号（突出“单个字符”）---*/
  for (auto [symbol, type] : std::initializer_list<std::pair<char, TokenType>>{
         {'>', TokenType::GreaterThan},
         {'<', TokenType::LessThan},
         {'+', TokenType::Plus},
         {'-', TokenType::Minus},
         {'*', TokenType::Asterisk},
         {'/', TokenType::Slash},
         {':', TokenType::Colon},
         {'=', TokenType::Equal},
         {',', TokenType::Comma},
         {';', TokenType::Semicolon},
         {'(', TokenType::LeftParenthesis},
         {')', TokenType::RightParenthesis},
         {'{', TokenType::LeftCurlyBrace},
         {'}', TokenType::RightCurlyBrace},
         {'[', TokenType::LeftSquareBracket},
         {']', TokenType::RightSquareBracket},
         {'.', TokenType::Dot},
         {'|', TokenType::Pipe},
       }) {
    lexer->register_machine(
      Machines::create_single_symbol_machine(symbol, type)
    );
  }

  /*--- 标识符---*/
  lexer->register_machine(Machines::create_identifier_machine());

  /*--- 空白字符---*/
  lexer->register_machine(Machines::create_whitespace_machine());
  lexer->register_machine(Machines::create_comment_machine());
  lexer->register_machine(Machines::create_newline_machine());
  lexer->register_machine(Machines::create_tab_machine());

  return lexer;
}

}  // namespace Lexer

File: next_kaubo/src/compiler/builder.rs
use super::token_kind::KauboTokenKind;
use crate::kit::lexer::{
    c_lexer::Lexer,
    state_machine::builder::{
        build_comment_machine, build_identifier_machine, build_integer_machine,
        build_keyword_machine, build_multi_char_machine, build_newline_machine,
        build_single_char_machine, build_string_machine, build_tab_machine,
        build_whitespace_machine,
    },
};
pub fn build_lexer() -> Lexer<KauboTokenKind> {
    let mut lexer: Lexer<KauboTokenKind> = Lexer::new(1024);

    for (kind, operator) in vec![
        (KauboTokenKind::DoubleEqual, "=="),
        (KauboTokenKind::LessThanEqual, "<="),
        (KauboTokenKind::GreaterThanEqual, ">="),
        (KauboTokenKind::ExclamationEqual, "!="),
    ] {
        lexer.register_machine(build_multi_char_machine(kind, operator.chars().collect()).unwrap());
    }

    for (kind, operator) in vec![
        (KauboTokenKind::GreaterThan, ">"),
        (KauboTokenKind::LessThan, "<"),
        (KauboTokenKind::Plus, "+"),
        (KauboTokenKind::Minus, "-"),
        (KauboTokenKind::Asterisk, "*"),
        (KauboTokenKind::Slash, "/"),
        (KauboTokenKind::Colon, ":"),
        (KauboTokenKind::Comma, ","),
        (KauboTokenKind::Semicolon, ";"),
        (KauboTokenKind::LeftParenthesis, "("),
        (KauboTokenKind::RightParenthesis, ")"),
        (KauboTokenKind::LeftCurlyBrace, "{"),
        (KauboTokenKind::RightCurlyBrace, "}"),
        (KauboTokenKind::LeftSquareBracket, "["),
        (KauboTokenKind::RightSquareBracket, "]"),
        (KauboTokenKind::Dot, "."),
        (KauboTokenKind::Pipe, "|"),
        (KauboTokenKind::Equal, "="),
    ] {
        lexer.register_machine(
            build_single_char_machine(kind, operator.chars().next().unwrap()).unwrap(),
        );
    }

    for (kind, keyword) in vec![
        (KauboTokenKind::Var, "var"),
        (KauboTokenKind::If, "if"),
        (KauboTokenKind::Else, "else"),
        (KauboTokenKind::Elif, "elif"),
        (KauboTokenKind::While, "while"),
        (KauboTokenKind::For, "for"),
        (KauboTokenKind::Return, "return"),
        (KauboTokenKind::In, "in"),
        (KauboTokenKind::Yield, "yield"),
        (KauboTokenKind::True, "true"),
        (KauboTokenKind::False, "false"),
        (KauboTokenKind::Null, "null"),
        (KauboTokenKind::Break, "break"),
        (KauboTokenKind::Continue, "continue"),
        (KauboTokenKind::Struct, "struct"),
        (KauboTokenKind::Interface, "interface"),
        (KauboTokenKind::Import, "import"),
        (KauboTokenKind::As, "as"),
        (KauboTokenKind::From, "from"),
        (KauboTokenKind::Pass, "pass"),
        (KauboTokenKind::And, "and"),
        (KauboTokenKind::Or, "or"),
        (KauboTokenKind::Not, "not"),
        (KauboTokenKind::Async, "async"),
        (KauboTokenKind::Await, "await"),
    ] {
        lexer.register_machine(build_keyword_machine(keyword, kind).unwrap());
    }

    lexer.register_machine(build_integer_machine(KauboTokenKind::LiteralInteger).unwrap());
    lexer.register_machine(build_string_machine(KauboTokenKind::LiteralString).unwrap());
    lexer.register_machine(build_identifier_machine(KauboTokenKind::Identifier).unwrap());

    lexer.register_machine(build_newline_machine(KauboTokenKind::NewLine).unwrap());
    lexer.register_machine(build_whitespace_machine(KauboTokenKind::Whitespace).unwrap());
    lexer.register_machine(build_tab_machine(KauboTokenKind::Tab).unwrap());
    lexer.register_machine(build_comment_machine(KauboTokenKind::Comment).unwrap());

    return lexer;
}
